csds-client/client/v3/client.go

```diff
@@ -0,0 +1,333 @@
+// Package client/v3 implements the client interface for v3 transport api version
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"envoy-tools/csds-client/client"
+	clientutil "envoy-tools/csds-client/client/util"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	csdspb_v3 "github.com/envoyproxy/go-control-plane/envoy/service/status/v3"
+	envoy_type_matcher_v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+// ClientV3 implements the Client interface
+type ClientV3 struct {
+	clientConn *grpc.ClientConn
+	csdsClient csdspb_v3.ClientStatusDiscoveryServiceClient
+
+	nodeMatcher []*envoy_type_matcher_v3.NodeMatcher
+	metadata    metadata.MD
+	opts        client.ClientOptions
+}
+
+// Field keys that must be presented in the NodeMatcher
+const (
+	gcpProjectNumberKey string = "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER"
+	gcpNetworkNameKey   string = "TRAFFICDIRECTOR_NETWORK_NAME"
+)
+
+// parseNodeMatcher parses the csds request yaml from -request_file and -request_yaml to nodematcher
+// if -request_file and -request_yaml are both set, the values in this yaml string will override and
+// merge with the request loaded from -request_file
+func (c *ClientV3) parseNodeMatcher() error {
+	if c.opts.RequestFile == "" && c.opts.RequestYaml == "" {
+		return errors.New("missing request yaml")
+	}
+
+	var nodematchers []*envoy_type_matcher_v3.NodeMatcher
+	if err := parseYaml(c.opts.RequestFile, c.opts.RequestYaml, &nodematchers); err != nil {
+		return err
+	}
+
+	c.nodeMatcher = nodematchers
+
+	// check if required fields exist in NodeMatcher
+	switch c.opts.Platform {
+	case "gcp":
+		keys := []string{gcpProjectNumberKey, gcpNetworkNameKey}
+		for _, key := range keys {
+			if value := getValueByKeyFromNodeMatcher(c.nodeMatcher, key); value == "" {
+				return fmt.Errorf("missing field %v in NodeMatcher", key)
+			}
+		}
+	default:
+		return fmt.Errorf("%s platform is not supported, list of supported platforms: gcp", c.opts.Platform)
+	}
+
+	return nil
+}
+
+// connWithAuth connects to uri with authentication
+func (c *ClientV3) connWithAuth() error {
+	var err error
+	switch c.opts.AuthnMode {
+	case "jwt":
+		switch c.opts.Platform {
+		case "gcp":
+			c.clientConn, err = clientutil.ConnToGCPWithJwt(c.opts.Jwt, c.opts.Uri)
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return fmt.Errorf("%s platform is not supported, list of supported platforms: gcp", c.opts.Platform)
+		}
+
+	case "auto":
+		switch c.opts.Platform {
+		case "gcp":
+			// parse GCP project number as header for authentication
+			projectNum := getValueByKeyFromNodeMatcher(c.nodeMatcher, gcpProjectNumberKey)
```

Comment:
You can inline this and next lines, it is more go-ish:

if p := getValueByKeyFromNodeMatcher(c.nodeMatcher, gcpProjectNumberKey);  p != "" {}